### ISSUE TITLE
New headless chrome randomly hangs at the end

### ISSUE THREAD

**Issue by maciejmrozinski** [cypress-io/cypress#27264]
### Current behavior

Using new headless mode in Chrome it sometimes hangs after running test file (also mentioned in this comment https://github.com/cypress-io/cypress/issues/25972#issuecomment-1487021361)

### Desired behavior

Consistent, correct behaviour just like with headless=old

### Test code to reproduce

There is no simple repo to reproduce. From my observation it hangs mostly if test flow contains the page that has external (iframe) loaded content (in my case it's google recaptcha). CI runs are more vulnerable to this issue, but I was able to reproduce it few times locally inside docker container.

Simple test code that sometimes fails:
```js
describe('test_1', () => {
    it('Test test', () => {
        cy.visit('/login');// visit page that has reCaptcha loaded
        cy.get('img.deskop__logo').click();// click on main logo, get back to homepage
    });
});
```
### Cypress Version

12.17.0

### Node version

18.16.0

### Operating System

Docker image: cypress/browsers:node-18.16.0-chrome-113.0.5672.92-1-ff-113.0-edge-113.0.1774.35-1

### Debug Logs

```shell
Last few lines of good run (exited correctly):
  cypress:launcher:browsers chrome stderr: [0711/114213.126568:ERROR:nacl_helper_linux.cc(355)] NaCl helper process running without a sandbox!
Most likely you need to configure your SUID sandbox correctly +831ms
  cypress:launcher:browsers chrome exited: { code: 0, signal: null } +1ms
  cypress:server:preprocessor removeFile /tests/files/test.ts +13s
  cypress:server:preprocessor base emitter plugin close event +0ms
  cypress:server:preprocessor base emitter native close event +2ms
  cypress:server:preprocessor base emitter native close event +0ms
  cypress:server:browsers:chrome closing remote interface client +416ms
  cypress:server:cypress about to exit with code 0 +19s
  cypress:webpack close /tests/files/test.ts +8s
  cypress:server:browsers browsers.kill called with no active instance +15s
  cypress:proxy:http:util:prerequests metrics: { browserPreRequestsReceived: 108, proxyRequestsReceived: 84, immediatelyMatchedRequests: 29, unmatchedRequests: 6, unmatchedPreRequests: 0 } +0ms
  cypress:cli child event fired { event: 'exit', code: 0, signal: null } +20s
  cypress:cli Stopping Xvfb +21s
  cypress:cli child event fired { event: 'close', code: 0, signal: null } +5ms

Last few lines of bad run (hangs):
  cypress:launcher:browsers chrome exited: { code: 0, signal: null } +1s
  cypress:server:preprocessor removeFile /tests/files/test.ts +18s
  cypress:server:preprocessor base emitter plugin close event +0ms
  cypress:server:preprocessor base emitter native close event +1ms
  cypress:server:preprocessor base emitter native close event +0ms
  cypress:server:browsers:chrome closing remote interface client +10ms
  cypress:webpack close /tests/files/test.ts +11s
  cypress:launcher:browsers chrome stderr: [0711/114246.895419:ERROR:nacl_helper_linux.cc(355)] NaCl helper process running without a sandbox!
Most likely you need to configure your SUID sandbox correctly +4ms
```

**Comment by jennifer-shehane** [comment]
@maciejmrozinski headless=new is what is run in Cypress by default since 12.15.0, so you don't need to pass the workaround tp get the new headless. Does turning off headless=new resolve your issue? Pass the code below, let us know.

```js
setupNodeEvents: function setupNodeEvents(on, config) {
  on('before:browser:launch', (browser = {}, launchOptions) => {
    if (browser.name === 'chrome' && browser.isHeadless) {
        launchOptions.args.push('--headless');
    }
    return launchOptions;
  });
}
```

**Comment by maciejmrozinski** [comment]
Yes, I've already tested this before and using old headless mode is resolving the issue.

**Comment by jennifer-shehane** [comment]
@maciejmrozinski We'll need a test to run to reproduce this behavior fully. We haven't observed this behavior on our test suite, so there is something particular about your run that's causing the hang that we'd need to be able to run so we can track down the issue.

**Comment by mikejav** [comment]
It's also applicable to my project.
Thx guys for mentioning `--headless=old`. It works like a charm.

**Comment by tiehfood** [comment]
We have the same issue, will there be a fix in the future?

**Comment by vire** [comment]
I have this issue as well on CI (CircleCI) I get `yarn exited with code 1` 🤷 

When I try `cypress run --record false --browser chrome --headless=old`
with config: 
```
> yarn cypress --version
Cypress package version: 12.17.3
Cypress binary version: 12.17.3
Electron version: 21.0.0
Bundled Node version: 16.16.0
```

I get 
<img width="1224" alt="image" src="https://github.com/cypress-io/cypress/assets/4021908/cdcdc69a-f810-4582-9d7c-448a72f05b37">

**Comment by cheuk0324** [comment]
Is there update for this issue?

**Comment by MikeMcC399** [comment]
@vire 

> I have this issue as well on CI (CircleCI) I get `yarn exited with code 1` 🤷
> 
> When I try `cypress run --record false --browser chrome --headless=old` with config:

`--headless=old` is not to be used as a CLI parameter for Cypress. It is a command line flag for Google Chrome. The way to pass this argument is described in https://github.com/cypress-io/cypress/issues/27264#issuecomment-1632768735. The CLI argument for Cypress is --headless. (See [Cypress Guides > Command Line > Options](https://docs.cypress.io/guides/guides/command-line#Options)).

See also [Google Chrome Developer > Try out the new Headless](https://developer.chrome.com/articles/new-headless/#try-out-the-new-headless)


- Be aware however that there is a new issue with Google Chrome `117` which stops this working. See https://github.com/cypress-io/cypress-documentation/issues/5483 for details. The Chrome arguments `--headless` and `--headless=old` are supposed to be equivalent, and this worked from Google Chrome `112` through `116`. In Chrome `117` there is a bug which crashes Chrome when these arguments are passed.
Edit: Fixed in Google Chrome `117.0.5938.132`.

**Comment by bericp1** [comment]
We're also encountering this issue with our cypress runs in CI.

We were using the `--headless=old` workaround but as mentioned that's broken in latest chrome. Removing that argument from browser launch options brought back this hanging issue for us.

![](https://d.pr/i/M4izRG.png)

![](https://d.pr/i/d3wC0P.png)

We use parallelization via DeploySentinel. You can see that one of the parallel runs in this example (#5) drops to 0% CPU and hangs, producing now output, eventually being cancelled by CircleCI due to inactivity.

I'll enable verbose debug logging on the process handler and the chrome browser launch to get chrom stderr and report back if that reveals anything.

**Comment by jennifer-shehane** [comment]
We need a reproducible example provided so that we can narrow down the cause of hanging for some users with the new headless behavior. Please can someone provide one.

**Comment by bericp1** [comment]
The problem is that it's not reproducible, it's seemingly random. We're not sure what triggers it.

**Comment by tiehfood** [comment]
We also still have this issue. Currently no clue on how to reproduce it, as it occurs randomly

**Comment by cheuk0324** [comment]
This issue is random, it seems it is hanging with one of the longest test or with the last instance. But this seems to get resolved after upgraded to the latest

**Comment by bericp1** [comment]
@jennifer-shehane we don't have a reproducible example but we did manage to capture debug logs for a run that this happened to. See attached file.

[circleci.com_api_v1.1_project_github_nursefly_nursefly-web_482932_output_120_4_file=true&allocation-id=650cecab5769582ef4129e07-4-build%2FABCDEFGH.txt](https://github.com/cypress-io/cypress/files/12702190/circleci.com_api_v1.1_project_github_nursefly_nursefly-web_482932_output_120_4_file.true.allocation-id.650cecab5769582ef4129e07-4-build.2FABCDEFGH.txt)

This is with the following:

```
DEBUG=cypress:server:util:process_profiler,cypress:launcher:browsers
```

The logs that look promising there to me are shortly before the tests just hang we get:

```
  [36;1mcypress:launcher:browsers [0mchrome exited: { code: [33m0[39m, signal: [1mnull[22m } [36m+3s[0m
  [36;1mcypress:launcher:browsers [0mchrome stderr: [0922/014838.655632:ERROR:nacl_helper_linux.cc(354)] NaCl helper process running without a sandbox!
```

And then the chrome process is missing from the process list that immediately follows. It pops up again later but maybe cypress isn't reconnecting to the new chrome process and hence the hanging?

**Comment by jsotelo** [comment]
`--headless=old` also worked for us. 

We did notice that peak memory usage was around 3.5 GB with new, whereas old uses about 2.4 GB. Our github actions runner has a 4GB resource limit. Perhaps chrome is running out of memory and is making cypress hang (pure guess).

We are using the following github actions config:
```yaml
  test:
    runs-on: [self-hosted, prod]
    container:
      image: cypress/included:cypress-13.2.0-node-20.6.1-chrome-116.0.5845.187-1-ff-117.0-edge-116.0.1938.76-1
      options: --ipc=host
```
and the following cypress.config.ts:
```ts
  e2e: {
    setupNodeEvents(_on, _config) {
      _on('before:browser:launch', (browser, launchOptions) => {
        if (browser.name === 'chrome') {
          launchOptions.args.push('--disable-dev-shm-usage');
          launchOptions.args.push('--headless=old');
        }
        console.log(launchOptions.args);
        return launchOptions;
      });
    },
```

**Comment by PavanGurram-DevOps** [comment]
Hi there, I'm also facing same issue with cypress version 12.7.4 and chrome version 112 when try to execute the tests in parallel. I have tried using above '--headless=old' but no luck.

Please can someone help? Thanks

**Comment by cheuk0324** [comment]
Our issue was resolved by itself after upgrade to v13

**Comment by PavanGurram-DevOps** [comment]
Unfortunately, I can't upgrade to v13 so need some workaround please

**Comment by jennifer-shehane** [comment]
We have observed a slowdown in performance for one project when using headless=new. We're still interested in having examples that show this behavior so that we can narrow down the issue. We suspect there's likely a bug in Chrome headless, but it's specific to some situation.

**Comment by joergschiller** [comment]
Not sure if it's really helpful but it could support the hypothesis that it's a bug in Chrome headless.

We're having the same issue that Chrome with new headless modes just hangs randomly after running all tests. But with a whole different stack: We're on Ruby and using RSpec/Capybara.

**Comment by Javediqbal2** [comment]
@jennifer-shehane I'm facing same issue with electron browser. Cypress tests hangs up sometimes at the end and sometimes before starting. In cypress cloud it show "This spec does not have any test results because it timed out". I've faced same issue from cypress 12.13.0 to 12.17.0 and for some people it hangs in firefox too. I'm mentioning this issue for reference 

https://github.com/cypress-io/github-action/issues/620

**Comment by jennifer-shehane** [comment]
Is this still occuring for people? We haven't had comments for a couple of months.

**Comment by pirate** [comment]
Yes, myself and other users of my project are still seeing headless chrome randomly hang before exit, even when it's run directly via CLI outside of cypress. I'm almost positive it's an upstream chrome bug. Rebooting often fixes it, waiting an hour and trying again sometimes fixes it, force-reinstalling chrome also often fixes it, which puts it squarely into heisenbug territory.

Related:
- https://github.com/ArchiveBox/ArchiveBox/issues/1045#issuecomment-1966605982
- https://github.com/ArchiveBox/ArchiveBox/issues/1125#issuecomment-1936600840

**Comment by jennifer-shehane** [comment]
@pirate What version of Chrome are you using? Have you tried updating?

**Comment by pirate** [comment]
This issue has been present as far back as `v60` but got much worse in `v112` (when we switched to the new `headless=new`), and has persisted all the way up to `v121.0.6167.57` and beyond with some versions worse than others.

It's intermittent and hard to verify sometimes, so many issues I've found about it on related projects have gotten closed as "cannot reproduce". I've just confirmed it's happening particularly consistently with `v121` though, but I still can't figure out why or when, as sometimes weird things like rebooting make it go away. I can post back here as I collect more reports on the latest versions.

There are also widespread reports of similar issues with the two telltale symptoms:
- `chrome headless seeems to hang indefinitely on exit sometimes`
- `my [insert headless driver here] appears to have a memory leak` (caused by chrome child processes hanging on exit and not releasing their memory)

Possibly related reports:
- https://support.google.com/chrome/thread/179100154/mac-os-chrome-hangs-on-exit?hl=en
- https://issues.chromium.org/issues/327458826
- https://github.com/puppeteer/puppeteer/issues/7922
- https://github.com/puppeteer/puppeteer/issues/1825
- https://forum.uipath.com/t/google-chrome-hangs-freezes-after-few-100-records/581107
- https://github.com/microsoft/playwright/issues/5327
- https://github.com/microsoft/playwright/issues/4218
- https://github.com/microsoft/playwright-python/issues/1074
- https://github.com/microsoft/playwright/issues/6319
- https://github.com/microsoft/playwright/issues/15400 (longstanding playwright issue caused by old chrome contexts not exiting properly and eating up memory, closed prematurely as you can see there are still recent comments about it happening)

It's possible some of these issues ^ are unrelated, but it's also possible they all stem from the same underlying issue of child chromium processes not exiting correctly.

The problem is widespread enough that many of the tools that use chrome headless have implemented hacky workarounds like this: https://devforth.io/blog/how-to-simply-workaround-ram-leaking-libraries-like-puppeteer-universal-way-to-fix-ram-leaks-once-and-forever/ (spawning chrome under a child process then doing killasgroup -9 after every run)

---

I tried again just for fun and managed to reproduce this on the first try!

I didn't even add any of the extra args we usually use (`--disable-gpu`, `--no-sandbox`, `--disable-features=dbus`, etc.), it hung immediately on the first try with only `--headless=new` and `--screenshot`!

<img width="1410" alt="Screenshot 2024-02-29 at 5 30 24 PM" src="https://github.com/ArchiveBox/ArchiveBox/assets/511499/5d65c33b-3c59-44e7-9ddb-12a8fc6bae9c">

This dispelled the last of my doubts, I think this is 100% an upstream Chromium bug and has nothing to do with Cypress/Playwright/Puppeteer/ArchiveBox/any driver.

**I just opened an upstream bug report on the Chromium bug tracker, follow over there for progress:** https://issues.chromium.org/issues/327583144 👾

<img width="600" alt="htop" src="https://github.com/cypress-io/cypress/assets/511499/431e9448-015d-481f-9535-1527ac626d87">

<img width="300" alt="profiling1" src="https://github.com/cypress-io/cypress/assets/511499/2804bbaf-406a-43a3-b241-f35b49d5f992"><img width="300" alt="profiling2" src="https://github.com/cypress-io/cypress/assets/511499/d418c87f-9978-4efd-b914-2c03123d6b10">

**Comment by jennifer-shehane** [comment]
@pirate Thanks for the detailed writeup and opening an issue with chromium. We'll take a look. It is extremely difficult to track down with all the variables involved as you explained. Is there a way to provide the project you're running where you got it to hang immediately?

**Comment by pirate** [comment]
It's not in any project, it's just raw chromium headless from the command line, no extra env vars, hidden CLI flags, or profile directory provided:
```bash
# using chromium downloaded via puppeteer
# recommended by: https://www.chromium.org/getting-involved/download-chromium/
$ npx @puppeteer/browsers install chrome@121.0.6167.57
$ ~/chrome/mac_arm-121.0.6167.57/chrome-mac-arm64/Google\ Chrome\ for\ Testing.app/Contents/MacOS/Google\ Chrome\ for\ Testing --headless=new --screenshot 'https://example.com'
[63086:259:0301/184829.306692:ERROR:policy_logger.cc(156)] :components/enterprise/browser/controller/chrome_browser_cloud_management_controller.cc(161) Cloud management controller initialization aborted as CBCM is not enabled. Please use the `--enable-chrome-browser-cloud-management` command line flag to enable it if you are not using the official Google Chrome build.
72602 bytes written to file screenshot.png
# ... hangs indefinitely ...
^C⏎
[138.101s]

# OR equivalent using playwright's chromium
$ pip install --upgrade playwright
$ playwright install --with-deps chromium
$ ~/Library/Caches/ms-playwright/chromium-1097/chrome-mac/Chromium.app/Contents/MacOS/Chromium --headless=new --screenshot 'https://example.com'
[63478:259:0301/185212.347544:ERROR:policy_logger.cc(156)] :components/enterprise/browser/controller/chrome_browser_cloud_management_controller.cc(161) Cloud management controller initialization aborted as CBCM is not enabled. Please use the `--enable-chrome-browser-cloud-management` command line flag to enable it if you are not using the official Google Chrome build.
72602 bytes written to file screenshot.png
# ... hangs indefinitely ...
^C⏎
[241.309s]
```
![both_puppeteer_and_playwright_chrome_hanging](https://github.com/cypress-io/cypress/assets/511499/c3ad4bbb-8198-40ed-a000-963d3d180d04)


Hung on the first try for both methods. Other versions besides `121.0.6167.57` do it too, but this one appears to do it particularly consistently on my machine. I can also reproduce this on freshly installed `Ubuntu 22.04`, and on both `x86` and `arm64` machines with both macOS and Linux.

**Comment by pirate** [comment]
@cypress-app-bot reopen

🙏

**Comment by pirate** [comment]
The chromium team is now looking at my bug report finally! https://issues.chromium.org/issues/327583144

![Image](https://github.com/user-attachments/assets/ff11245e-2173-488a-8b49-65a6b52f1c20)
